scanner/updater/export.go

```diff
@@ -0,0 +1,166 @@
+package updater
+
+import (
+	"compress/gzip"
+	"context"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/quay/claircore/alpine"
+	"github.com/quay/claircore/aws"
+	"github.com/quay/claircore/debian"
+	"github.com/quay/claircore/libvuln/driver"
+	"github.com/quay/claircore/libvuln/jsonblob"
+	"github.com/quay/claircore/libvuln/updates"
+	"github.com/quay/claircore/oracle"
+	"github.com/quay/claircore/photon"
+	"github.com/quay/claircore/rhel"
+	"github.com/quay/claircore/suse"
+	"github.com/quay/claircore/ubuntu"
+	"github.com/quay/claircore/updater/osv"
+	"github.com/quay/zlog"
+	"github.com/stackrox/stackrox/scanner/v4/updater/manual"
+)
+
+// Export is responsible for triggering the updaters to download Common Vulnerabilities and Exposures (CVEs) data
+// and then outputting the result as a gzip file
+func Export(ctx context.Context) error {
+	updaterStore, err := jsonblob.New()
+	if err != nil {
+		return err
+	}
+
+	var updaters []driver.Updater
+
+	// Append updater sets directly to the updaters
+	appendUpdaterSet := func(updaterSet driver.UpdaterSet, err error) {
+		if err != nil {
+			zlog.Error(ctx).Msg(err.Error())
+			return
+		}
+		updaters = append(updaters, updaterSet.Updaters()...)
+	}
+
+	appendUpdaterSet(aws.UpdaterSet(ctx))
+	appendUpdaterSet(manual.UpdaterSet(ctx, nil))
+	appendUpdaterSet(oracle.UpdaterSet(ctx))
+	appendUpdaterSet(photon.UpdaterSet(ctx))
+	appendUpdaterSet(suse.UpdaterSet(ctx))
+
+	alpineFac, err := alpine.NewFactory(ctx)
+	if err != nil {
+		return err
+	}
+
+	debianFac, err := debian.NewFactory(ctx)
+	if err != nil {
+		return err
+	}
+
+	rhelFac, err := rhel.NewFactory(ctx, rhel.DefaultManifest)
+	if err != nil {
+		return err
+	}
+
+	ubuntuFac, err := ubuntu.NewFactory(ctx)
+	if err != nil {
+		return err
+	}
+
+	osvFac := osv.Factory
+
+	cfgs := map[string]driver.ConfigUnmarshaler{
+		"debian": func(v interface{}) error {
+			v.(*debian.FactoryConfig).MirrorURL = `https://deb.debian.org/`
+			return nil
+		},
+		"alpine": func(v interface{}) error {
+			v.(*alpine.FactoryConfig).URL = "https://secdb.alpinelinux.org/"
+			return nil
+		},
+		"rhel": func(v interface{}) error {
+			return nil
+		},
+		"ubuntu": func(v interface{}) error {
+			v.(*ubuntu.FactoryConfig).Name = "ubuntu"
+			return nil
+		},
+		"osv": func(v interface{}) error {
+			cfg := v.(*osv.FactoryConfig)
+			cfg.URL = osv.DefaultURL
+			return nil
+		},
+	}
+
+	facs := map[string]driver.UpdaterSetFactory{
+		"debian": debianFac,
+		"alpine": alpineFac,
+		"rhel":   rhelFac,
+		"ubuntu": ubuntuFac,
+		"osv":    osvFac,
+	}
+
+	os.Mkdir("tmp", 0700)
```

Comment:
Can we allow the caller to choose storage and location for the bundle output? It would give them more control and options.

Reply:
This folder and the output content will be uploaded to google storage so I don't think it matters too much regarding the folder name and path.

Reply:
@daynewlee:

> This folder and the output content will be uploaded to google storage, so I don't think it matters too much regarding the folder name and path.

Even if the output directory primary is to be uploaded to Google Storage, allowing the tool to run based on a configurable directory makes sense. First, the updater should ideally be environment agnostic. It makes the code more testable and versatile, meaning it should work the same whether running on the developer's machine, UTs or in a CI/CD environment, and hard coding the directory is guaranteed to be safe only in the latter. The updater is intended to be used in a CI/CD environment, but requirements may change. Second, readability is poor with this hardcoded path. Reading either CI or the Updater code, you can't tell why specifically we are `mv tmp ${{ env.ROX_PRODUCT_VERSION }}`, where did `tmp` come from? Most likely, the reader will ask themselves: "Why did we hard code `tmp`? Is it a CI constraint?". Finally, the cost of accepting and passing a command line argument when running the command is extremely low to justify not doing it for simplicity or quick delivery.

Example:

```
mkdir  "${{ env.ROX_PRODUCT_VERSION }}"
go run cmd/updater/update.go "${{ env.ROX_PRODUCT_VERSION }}/vulnerability-dump.json.gz"
gsutil cp -r "${{ env.ROX_PRODUCT_VERSION }}" "gs://<GOOGLE-BUCKET>"
```

(Notice: I also changed the bundle name to "vulnerability-dump.json.gz" because "output.json.gz" doesn't tell me anything about what this file is, while the former does.)

TL;DR: The best practice is to avoid hardcoding the directory path, even in a CI/CD environment.

Reply:
Yeah, I think It could be useful to make it configurable if we want to run the updater locally for testing or some other reason

Reply:
> @daynewlee:
> 
> > This folder and the output content will be uploaded to google storage, so I don't think it matters too much regarding the folder name and path.
> 
> Even if the output directory is to be uploaded to Google Storage, allowing the tool to run based on a configurable path makes sense. First, the updater should ideally be environment agnostic. It makes the code more testable and versatile, meaning it should work the same whether running on the developer's machine, UTs or in a CI/CD environment, and hard coding the directory is guaranteed to be safe only in the latter. The updater is intended to be used in a CI/CD environment, but requirements may change. Second, readability is poor with this hardcoded path. Reading either CI or the Updater code, you can't tell why specifically we are `mv tmp ${{ env.ROX_PRODUCT_VERSION }}`, where did `tmp` come from? Most likely, the reader will ask themselves: "Why did we hard code `tmp`? Is it a CI constraint?". Finally, the cost of accepting and passing a command line argument when running the command is extremely low to justify not doing it for simplicity or quick delivery.
> 
> Example:
> 
> ```
> mkdir  "${{ env.ROX_PRODUCT_VERSION }}"
> go run cmd/updater/update.go "${{ env.ROX_PRODUCT_VERSION }}/vulnerability-dump.json.gz"
> gsutil cp -r "${{ env.ROX_PRODUCT_VERSION }}" "gs://<GOOGLE-BUCKET>"
> ```
> 
> (Notice: I also changed the bundle name to "vulnerability-dump.json.gz" because "output.json.gz" doesn't tell me anything about what this file is, while the former does.)
> 
> TL;DR: The best practice is to avoid hardcoding the directory path, even in a CI/CD environment.

This actually makes sense. We don't have to do the `mv tmp to xxx step`.